### ISSUE TITLE
boto raises BadStatusLine exception when using moto.emr and boto.emr.connect_to_region()

### ISSUE THREAD

**Issue by achiku** [spulec/moto#347]
Hi! First of all thank you so much for open sourcing this wonderful library!

I have a question related to ```moto.emr``` module. Is it possible to use ```boto.emr.connect_to_region()``` to get EmrConnection while using moto? 

What I wanted to do
-------------------
```boto.emr.connect_to_region()``` is used in my code, and I wanted to mock EMR in my test code.

Errors
------

When I run the test above, it raises BadStatusLine exception.

```
/Users/achiku/.virtualenvs/dashboard/src/boto/boto/emr/connection.py:580: in run_jobflow
    'RunJobFlow', params, RunJobFlowResponse, verb='POST')
/Users/achiku/.virtualenvs/dashboard/src/boto/boto/connection.py:1192: in get_object
    response = self.make_request(action, params, path, verb)
/Users/achiku/.virtualenvs/dashboard/src/boto/boto/connection.py:1116: in make_request
    return self._mexe(http_request)
/Users/achiku/.virtualenvs/dashboard/src/boto/boto/connection.py:1030: in _mexe
    raise ex
E   BadStatusLine: ''
```

Error Reproduction
------------------

moto EMR test cases are using ```boto.connect_emr()```, and when I clone moto repo, make init, and run the tests(make test), all tests successfully passed. Then I silightly modify ```tests/test_emr/test_emr.py``` and run the tests again to reproduce the error above.

```
$ python --version
Python 2.7.9
$ pip freeze
You are using pip version 6.0.8, however version 6.1.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
boto==2.38.0
coverage==3.7.1
Flask==0.10.1
freezegun==0.3.3
gnureadline==6.3.3
httpretty==0.8.9
ipython==3.1.0
itsdangerous==0.24
Jinja2==2.7.3
MarkupSafe==0.23
mock==1.0.1
-e git+git@github.com:spulec/moto.git@be5f0414160ca254d9c6df5e13917d52969c269b#egg=moto-master
nose==1.3.6
python-dateutil==2.4.2
requests==2.7.0
six==1.9.0
sure==1.2.3
Werkzeug==0.10.4
xmltodict==0.9.2
```

```
diff --git a/tests/test_emr/test_emr.py b/tests/test_emr/test_emr.py
index 9157a82..d6065b9 100644
--- a/tests/test_emr/test_emr.py
+++ b/tests/test_emr/test_emr.py
@@ -10,7 +10,7 @@ from tests.helpers import requires_boto_gte

 @mock_emr
 def test_create_job_flow():
-    conn = boto.connect_emr()
+    conn = boto.emr.connect_to_region(
+        'us-west-2', aws_access_key_id="ak", aws_secret_access_key="sk")

     step1 = StreamingStep(
         name='My wordcount example',
```

Test result is the following. It seems like the request retry count exceeds its limit.

```
$ nosetests -v tests/test_emr/test_emr.py
test_emr.test_add_steps_to_flow ... ok
test_emr.test_cluster_tagging ... ok
test_emr.test_create_instance_groups ... ok
test_emr.test_create_job_flow ... ERROR
test_emr.test_create_job_flow_visible_to_all_users ... ok
test_emr.test_create_job_flow_with_new_params ... ok
test_emr.test_describe_cluster ... ok
test_emr.test_describe_job_flows ... ok
test_emr.test_list_clusters ... ok
test_emr.test_modify_instance_groups ... ok
test_emr.test_set_visible_to_all_users ... ok
test_emr.test_terminate_job_flow ... ok

==================================================================
ERROR: test_emr.test_create_job_flow
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/achiku/.virtualenvs/moto/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/achiku/work/kanmu/sandbox_libs/moto/moto/core/models.py", line 70, in wrapper
    result = func(*args, **kwargs)
  File "/Users/achiku/work/kanmu/sandbox_libs/moto/tests/test_emr/test_emr.py", line 37, in test_create_job_flow
    steps=[step1, step2],
  File "/Users/achiku/.virtualenvs/moto/lib/python2.7/site-packages/boto/emr/connection.py", line 581, in run_jobflow
    'RunJobFlow', params, RunJobFlowResponse, verb='POST')
  File "/Users/achiku/.virtualenvs/moto/lib/python2.7/site-packages/boto/connection.py", line 1192, in get_object
    response = self.make_request(action, params, path, verb)
  File "/Users/achiku/.virtualenvs/moto/lib/python2.7/site-packages/boto/connection.py", line 1116, in make_request
    return self._mexe(http_request)
  File "/Users/achiku/.virtualenvs/moto/lib/python2.7/site-packages/boto/connection.py", line 1030, in _mexe
    raise ex
BadStatusLine: ''

----------------------------------------------------------------------
Ran 12 tests in 47.321s

FAILED (errors=1)
```

I woundered and looked into tests for other services, and found dynamodb is using connect_to_region().

- https://github.com/spulec/moto/blob/master/tests/test_dynamodb2/test_dynamodb.py#L43

If you have any idea to work around this exception, please point me out. I will look into it. Thanks!

**Comment by achiku** [comment]
It seems like moto.emr uses ```us-east-1``` as default region, and not supports other regions for now. I will work around this making region as environmental variable and use ```us-east-1``` in the local test.

**Comment by 8maki** [comment]
me too.

**Comment by spulec** [comment]
Making EMR support multi-region would be great. An example of how we accomplished that for SNS is https://github.com/spulec/moto/commit/53acdf6c763b87e6ed96553ebbd2dd705e98144d

**Comment by achiku** [comment]
Thanks for the comment! I'll send p-r as soon as it's ready!

**Comment by achiku** [comment]
umm, EMR seems like having different endpoints depending on regions. What boto's log tells us and the documentation seem different. I'll ask AWS support first.

- http://docs.aws.amazon.com/general/latest/gr/rande.html#emr_region


```
# -*- coding: utf-8 -*-
from datetime import datetime

import boto
from boto.emr import connect_to_region
from boto.emr.instance_group import InstanceGroup
from boto.emr.step import InstallHiveStep

_yourbucket = 'yourbucket'
_yoursshkey = 'yoursshkey'
_aws_access_key_id = 'yourkeyid'
_aws_secret_access_key = 'yoursecretkey'


boto.set_stream_logger('boto')

if __name__ == '__main__':
    bucket_name = _yourbucket
    install_hive = InstallHiveStep(hive_versions='0.13.1')

    setup_steps = [install_hive]
    instance_group = [
        InstanceGroup(
            1, 'MASTER', 'r3.xlarge',
            'SPOT', 'emr-master', bidprice="0.09"),
        InstanceGroup(
            2, 'CORE', 'r3.xlarge',
            'SPOT', 'emr-core', bidprice="0.09"),
    ]

    # this connection will log :Host: elasticmapreduce.us-east-1.amazonaws.com
    conn = connect_to_region(
        'us-east-1',
        aws_access_key_id=_aws_access_key_id,
        aws_secret_access_key=_aws_secret_access_key,
    )
    # but this logs :Host: us-west-1.elasticmapreduce.amazonaws.com
    # conn = connect_to_region(
    #     'us-west-1',
    #     aws_access_key_id=_aws_access_key_id,
    #     aws_secret_access_key=_aws_secret_access_key,
    # )
    jobid = conn.run_jobflow(
        name='boto_test_{}'.format(datetime.today().strftime('%Y%m%d%H%m%s')),
        ami_version='3.5.0',
        ec2_keyname=_yoursshkey,
        instance_groups=instance_group,
        action_on_failure='TERMINATE_JOB_FLOW',
        log_uri='s3://{0}/jobflow_logs/'.format(bucket_name),
        keep_alive=True,
        enable_debugging=False,
        steps=setup_steps,
        additional_info=None,
        api_params=None,
        visible_to_all_users=True,
        job_flow_role='EMR_EC2_DefaultRole',
        service_role='EMR_DefaultRole'
    )

    print jobid

```

**Comment by achiku** [comment]
EMR has a bit strange endpoint settings in boto although we can access both ```elasticmapreduce.<region>.amazonaws.com``` and  ```<region>.elasticmapreduce.amazonaws.com``` except ```eu-central-1``` and ```cn-north-1```. I don't know why boto is using this style only for EMR.

- https://github.com/boto/boto/blob/develop/boto/endpoints.json#L196

But I think I figured out to make mock url fit into this style.

**Comment by achiku** [comment]
I created p-r for boto repo. If this is merged, I will create p-r for multi-region emr module to moto repo.

https://github.com/achiku/moto/pull/1

**Comment by spulec** [comment]
Great, thanks for that. Keep me posted.

**Comment by achiku** [comment]
@spulec 
boto is not going to merge my p-r which makes EMR endpoints aligned with the official documentation since it breaks python 2.6 compatibility. 

https://github.com/boto/boto/pull/3186

Could I send p-r making moto EMR module multi-region aware? As you stated in this issue https://github.com/spulec/moto/issues/348, it's supposed to work and adding multi-region functionality to EMR module should not break current version of moto.

**Comment by spulec** [comment]
A PR to add multi-region support to EMR would be great!

**Comment by spulec** [comment]
Closed with #375
